3rdparty/everest/library/Hacl_Curve25519_joined.c

```diff
@@ -19,7 +19,7 @@
  *  This file is part of mbed TLS (https://tls.mbed.org)
  */
 
-#include "common.h"
+#include "internal/common.h"
```

Comment:
As I wrote in https://github.com/ARMmbed/mbedtls/issues/4032#issuecomment-782201943, I don't think we should move headers from `library` to `include/internal`. I see two practical reasons:

* It would make sense to have a directory called `include/mbedtls/internal` containing header files which have to be published for the sake of the C compiler, but that application code should not depend on. (For example: struct definitions where we want the compiler to know the size, but we don't want application code to access fields directly.) Having `include/internal` with a different meaning is confusing.
* Installers can currently copy `include/*` to the system include path. `include/internal` breaks this.

Reply:
Right sorry, I misunderstood and thought you were only referring to moving source files in that comment.

I agree with your above statement that `include/mbedtls/internal` would be a more suitable place for the internal headers, especially as it would presumably also allow us to split the PSA headers into `include/psa/internal`.

In your comment above it sounds like you would prefer to keep some header files out of `include/<mbedtls/psa>/internal` altogether. Can you clarify whether that is the case and if so where should headers that don't belong be put? For example: `ssl_invasive.h` looks like it might not fit into `include/mbedtls/internal` however I also don't agree with it being left in `library`, which at least to me seems like it should only contain source files (although I often come from a background of having a `src/` folder).

Reply:
Right, thanks for the clarification there.

Upon re-thinking I'm not against having headers inside `library/`, my main concern is unifying them in one location or at least having them in some sort of logical order. When I first cloned the library my first thought was that all the headers will be in `include` and the source files in `library/`. To me it seems strange and confusing that there are internal headers **and** public headers in `include/` **as well as** internal headers inside `library/`. This is made doubly so when the headers inside `library/` are are not named `*_internal.h` or have any mention of being only internal in the file itself (e.g: `ssl_tls13_keys.h`).

I don't think I have a particularly strong opinion on exactly where in `library/` they all go (I slightly prefer `library/<internal/include>` and then `library/*_internal.h`) but at the very least it should be documented and obvious that the files are part of the internal library and should not be used by application code.

Reply:
Thanks for the input @mpg, I was not aware there was a change in how internal headers are done which explains the discrepancy. I feel as though this makes the case that this is something that should be better documented so there is less reliance on knowing the history of the library.

Reply:
> When I first cloned the library my first thought was that all the headers will be in include and the source files in library/.

I'm puzzled as to why you thought this. It isn't a particularly common convention to have all header files, both public and private, under a single toplevel directory. Looking at a few examples:

* BearSSL: public headers in `inc`, private headers in `src`
* OpenSSL: public headers in `include`, private headers under `crypto` and `ssl` (and others).
* GnuTLS: public headers in `lib/includes`, private headers in `lib` (outside `lib/includes`) and `src` and others
* WolfSSL: no toplevel directory for include files, at first glance it looks like all headers in various library subdirectories are public and there are no private headers.
* Libsodium: public headers in `src/libsodium/include`, private headers in various other subdirectories of `lib/libsodium`.
* Curl: public headers in `include`, private headers in `lib` and `src`.

All of these include header files in the same directory as `.c` files. So I dispute the assertion that there's anything surprising about having header files in `library`.

> To me it seems strange and confusing that there are internal headers and public headers in include/ as well as internal headers inside library/.

But that's precisely what we wanted to fix! By moving the internal headers out of `include`. And into `library`.

Reply:
Thanks Gilles for making this list. Regarding the other point of discussion (whether putting a header outside of `include` is enough to identify it as internal) I checked BearSSL, OpenSSL, GnuTLS, libsodium, and none of them felt the need to add `_internal` or anything else to the names of headers outside of the public include directory in order to mark them as internal: the fact that they're out of the pubic include directory is enough. (Generally speaking, I'm sure I've see this pattern in other C projects, so to me at least that seems like a widespread and unsurprising convention.)

Coming back to an earlier point in the discussion, there are in principle 3 kinds of headers:

1. Fully public headers: application code can use what they declare.
2. Semi public headers: the compiler needs to see this when building the application, but application developers should not use that's declared there.
3. Private headers: the compiler only needs to see this when building the library, and only the library should use what's declared there.

As Gilles said, only fully and semi public headers have any business being in `include` (as long as `make install` copies all of it). Currently we only have fully public and private headers. As Gilles said, if we ever have semi-public headers, I think their natural place is in `include/mbedtls/internal` or any other place under `include` with `internal` in the path. (Btw, OpenSSL does something similar here: it has `include/openssl` for fully public headers and `include/internal` for semi-public ones.)

Now regarding private headers, before this PR we had two groups:

- 3a. those in `include/mbedtls`, most often with `_internal` in the name but not always (see `bn_mul.h`), whose location+name would perhaps be appropriate for (semi-)public headers
- 3b. those in `library`, with nothing special in the name.

The goal of this PR, that's a point we all agree on, is to only have one group of private headers.

It was and still is my opinion that the most simple and best way to achieve that is to move 3a to 3b. Instead you felt the need to move both 3a and 3b to a new 3c "library/*_internal.h". I don't see a need for that, I think 3b is the most common convention and that it's a good one.

Reply:
Internal headers have been unified under `library/`. Would you approve this design now?